src/core/NewDomain.ml

```diff
@@ -0,0 +1,3788 @@
+    match data_tyx with
+    | Data datax ->
+
+      let rec go tyenv tele args : constr_cell list =
+        match tele, args with
+        | Desc.TNil _, [] -> []
+        | Desc.TCons (`Const tty, Tm.B (_, tele)), `Const v :: args ->
+          let tyx = Syn.eval rel tyenv tty in
+          let coe_hd s =
+            let x', pi = Perm.freshen_name x in
+            let tyx' = Con.swap pi tyx in
+            make_coe rel r s ~abs:(Abs (x', tyx')) v
+          in
+          let coe_tl =
+            let coe_hd_x = coe_hd @@ `Atom x in
+            let tyenv = Env.extend_cell tyenv @@ Cell.con coe_hd_x in
+            go tyenv tele args
+          in
+          `Const (Val.make @@ coe_hd r') :: coe_tl
+        | Desc.TCons (`Rec Desc.Self, Tm.B (_, tele)), `Rec (`Self, v) :: args ->
+          let coe_hd = rigid_coe rel r r' ~abs:data_abs v in
+          let tyenv = Env.extend_cell tyenv @@ Cell.con coe_hd in
+          let coe_tl = go tyenv tele args in
+          `Rec (`Self, Val.make coe_hd) :: coe_tl
+        | Desc.TCons (`Dim, Tm.B (_, tele)), `Dim s :: args ->
+          let tyenv = Env.extend_cell tyenv @@ Cell.dim s in
+          `Dim s :: go tyenv tele args
+        | _ ->
+          raise PleaseRaiseProperError
+
+      in
+
+      let genv, constrs = datax.constrs in
+      let constr = Desc.lookup_constr clbl constrs in
+      let tyenv = Env.extend_cells (Env.init genv) datax.params in
+      go tyenv constr args
+
+    | _ ->
+      raise PleaseRaiseProperError
+
+  and multi_coe rel r r' data_abs clbl args =
+    match Rel.equate r r' rel with
+    | `Same ->
+      args
+    | _ ->
+      rigid_multi_coe rel r r' data_abs clbl args
+
+  and constr_cell_to_cell : constr_cell -> cell =
+    function
+    | `Const v -> Cell.value v
+    | `Rec (_, v) -> Cell.value v
+    | `Dim r -> Cell.dim r
+
+  (* TODO: pass the data-info pre-extracted *)
+  (** invariant: the binder in [abs] {b must} be fresh so that we do not have to
+      hide the bound variable from [rel] or worry about shadowing. *)
+  and rigid_coe_nonstrict_data rel r r' ~abs cap =
+    let Abs (x, tyx) = abs in
+    match tyx, Val.unleash cap with
+    | Data data, Intro intro ->
+      let genv, constrs = data.constrs in
+      let constr = Desc.lookup_constr intro.clbl constrs in
+      let coerced_args s s' = multi_coe rel s s' abs intro.clbl intro.args in
+      let args_rr' = coerced_args r r' in
+      let tboundary = Desc.Constr.boundary constr in
+      let env_ps = Env.extend_cells (Env.init genv) data.params in
+      let intro =
+        let benv = Env.extend_cells env_ps @@ List.map constr_cell_to_cell args_rr' in
+        let sys = Syn.eval_tm_sys rel benv tboundary in
+        make_intro rel ~dlbl:data.lbl ~clbl:intro.clbl ~args:(coerced_args r r') ~sys
+      in
+      begin
+        match tboundary with
+        | [] -> intro
+        | _ ->
+          let faces =
+            let args_rx = coerced_args r @@ `Atom x in
+            let benv 
```

Comment:
The OCaml compiler may be unhappy, and I am also unhappy with `exception Triv` :wink:  I bet we'll find a lot of bugs if we get rid of such things elsewhere too.